src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -578,15 +620,24 @@ public void Add(T item, EntryState state)
                     }
                     else if (_currentState != state)
                     {
-                        _states = ArrayBuilder<EntryState>.GetInstance(_items.Count - 1, _currentState.Value);
+                        // Create a builder with the right capacity (so we don't waste scratch space). Copy all the same
+                        // prior values all the way up to the last item we're about to add.
+                        _states = ArrayBuilder<EntryState>.GetInstance(_requestedCapacity);
+                        for (int i = 0, n = _items.Count - 1; i < n; i++)
+                            _states.Add(_currentState.Value);
```

Comment:
We're not using `i` at all in here? What am I missing?

Reply:
Also, please add braces.

Reply:
We're filling all but the last item with the prior item we had seen. 

The idea of The type this is building is that it optimizes for the case that all the states are the same. I'm that case, it just uses a singleton, preallocated array with that value in it. 

So during building, if we keep seeing the same value, we do nothing but store that value. However, there moment we see a second value we actually have to generate the full array. So we generate where all the items but the last are the first values we saw, and the last item is the current different value we just ran into. 

Lmk if that helps.

Reply:
That does help, thank. I think a small note of that fact (that all previous states are the same if we're falling out of the optimized path) would be helpful.

Reply:
sure! happy to expand on the comment.  i agree it's subtle :)